test/cpp/end2end/tls_private_key_offload_end2end_test.cc

```diff
@@ -0,0 +1,422 @@
+  server_addr_ = absl::StrCat("localhost:",
+                              std::to_string(grpc_pick_unused_port_or_die()));
+  std::string server_key =
+      grpc_core::testing::GetFileContents(std::string(kServerKeyPath));
+  std::string server_cert =
+      grpc_core::testing::GetFileContents(std::string(kServerCertPath));
+  std::string ca_cert =
+      grpc_core::testing::GetFileContents(std::string(kCaPemPath));
+
+  experimental::IdentityKeyCertPair server_key_cert_pair;
+  server_key_cert_pair.private_key = server_key;
+  server_key_cert_pair.certificate_chain = server_cert;
+  std::vector<experimental::IdentityKeyCertPair> server_identity_key_cert_pairs;
+  server_identity_key_cert_pairs.emplace_back(server_key_cert_pair);
+  auto server_certificate_provider =
+      std::make_shared<experimental::StaticDataCertificateProvider>(
+          ca_cert, server_identity_key_cert_pairs);
+
+  absl::Notification notification;
+  server_thread_ = new std::thread(
+      [&]() { RunServer(&notification, server_certificate_provider); });
+  notification.WaitForNotification();
+
+  std::string client_key =
+      grpc_core::testing::GetFileContents(std::string(kClientKeyPath));
+  std::string client_cert =
+      grpc_core::testing::GetFileContents(std::string(kClientCertPath));
+  experimental::IdentityKeyCertPair client_key_cert_pair;
+  client_key_cert_pair.private_key = client_key;
+  client_key_cert_pair.certificate_chain = client_cert;
+  std::vector<experimental::IdentityKeyCertPair> client_identity_key_cert_pairs;
+  client_identity_key_cert_pairs.emplace_back(client_key_cert_pair);
+  auto client_certificate_provider =
+      std::make_shared<experimental::StaticDataCertificateProvider>(
+          ca_cert, client_identity_key_cert_pairs);
+  grpc::experimental::TlsChannelCredentialsOptions options;
+  options.set_certificate_provider(client_certificate_provider);
+  options.watch_root_certs();
+  options.set_root_cert_name("root");
+  options.watch_identity_key_cert_pairs();
+  options.set_identity_cert_name("identity");
+  options.set_check_call_host(false);
+
+  DoRpc(server_addr_, options);
+}
+
+TEST_F(TlsPrivateKeyOffloadTest, OffloadWithCustomKeySigner) {
+  server_addr_ = absl::StrCat("localhost:",
+                              std::to_string(grpc_pick_unused_port_or_die()));
+  std::string server_key =
+      grpc_core::testing::GetFileContents(std::string(kServerKeyPath));
+  std::string server_cert =
+      grpc_core::testing::GetFileContents(std::string(kServerCertPath));
+  std::string ca_cert =
+      grpc_core::testing::GetFileContents(std::string(kCaPemPath));
+
+  auto server_certificate_provider =
+      std::make_shared<experimental::InMemoryCertificateProvider>();
+  signer_ = std::make_shared<TestCustomPrivateKeySigner>(server_key);
+  grpc_core::PemKeyCertPairList identity_pairs;
+  identity_pairs.emplace_back(signer_, server_cert);
+  static_cast<grpc_core::InMemoryCertificateProvider*>(
```

Comment:
These tests should be using the public C++ API.  It should not need to reach inside to the C-core API.  The fact that you're needing to do this is a symptom of the problem that I pointed out elsewhere, which is that we are not exposing a way to set the private key signer in the C++ API.